site/content/blog/2020-02-10-maintaining-consistency-in-codebases-with-go-vet.md

```diff
@@ -0,0 +1,181 @@
+responsible for receiving the ast (and some other things), find the things that
+we consider errors, notify go vet of those errors, and alert the user.
+
+Let's build our Analyzer.
+
+```go
+// File: checkwords.go
+package main
+
+import (
+	"go/ast"
+	"go/token"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/unitchecker"
+)
+
+var analyzer = &analysis.Analyzer{
+	Name: "checkWords",
+	Doc:  "check forbidden words usage in our strings",
+	Run:  run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	forbiddenWords := []string{
+		"bird",
+		"water",
+		"candy",
+	}
+
+	for _, file := range pass.Files {
+		ast.Inspect(file, func(node ast.Node) bool {
+			switch x := node.(type) {
+			case *ast.BasicLit:
+				if x.Kind != token.STRING {
+					return false
+				}
+				words := strings.Fields(x.Value)
+				for _, word := range words {
+					for _, forbiddenWord := range forbiddenWords {
+						if word == forbiddenWord {
+							pass.Reportf(x.Pos(), "Forbidden word used, please avoid use the word %s on your strings", word)
+						}
+					}
+				}
+				return false
+			}
+			return true
+		})
+	}
+	return nil, nil
+}
+
+func main() {
+	unitchecker.Main(
+		analyzer,
+	)
+}
+```
+
+Our Analyzer is inspecting all the files searching for `*ast.BasicLit` of `Kind`
+`token.STRING`, which are our literal strings. It splits those strings by spaces,
+and checks whether any of them match the forbidden words (this is a completely
+basic approach, and doesn't catch a lot of cases, but for the sake of
+simplicity I'll leave it as is). If it finds any forbidden words, it reports to
+the user with an error message. Go vet handles printing the filename and
+location of the error.
+
+Once we have our Analyzer we only have to register the Analyzer in our main
+function to connect it with go vet using the `unitchecker.Main` function (we can
+register multiple analyzers there).
+
+Now we only need to compile it with `go build ./checkwords.go` and use it with
+`go vet -vettool=./checkwords -checkWords ./file-or-module-path`.
+
+For example, we can create an example.go file like this:
+
+```go
+package main
+
+import "fmt"
+
+func main() {
+	fmt.Println("my candy is forbidden!")
+	fmt.Println("but other strings are not")
+}
+```
+
+and run our vet tool to check this with `go vet -vettool=./checkwords -checkWords example.go` and the resulting output is:
+
+```
+# command-line-arguments
+./example.go:6:14: Forbidden word used, please avoid use the word candy on your strings
+```
+
+And that is all we need. Now we have an automatic way to detect undesired
+patterns in our code.
+
+Our conclusion is, using go vet is an excellent option to improve your code,
+but extending it allows you to define your own patterns and maintain the
+consistency of your code. If you see yourself asking for the same changes in
+the PRs all the time, you can probably consider detecting this kind of problem
+automatically with go vet.
```

Comment:
We are using our version for months already and our conclusion is using go vet tool is an excellent opportunity to improve your code, also, extending it allows you to define your own patterns and maintain the consistency of your code. With our open source culture you can find our implementations at our [mattermost-govet](https://github.com/mattermost/mattermost-govet) repository. If you see yourself asking for the same changes in the PRs all the time, you can probably consider detecting this kind of problem automatically with go vet.

Reply:
Suggestion for above:
"We have been using our version for a number of months and our conclusion is that using the go vet tool is an excellent opportunity to improve your code. In addition extending it allows you to define your own patterns and maintain the consistency of your code. With our open source culture you can find our implementations at our mattermost-govet repository. If you see yourself asking for the same changes in the PRs all the time, you can probably consider using go vet to detect the issue/s automatically."